### PR TITLE
[Enhancement] Convert fast-unstake to use StakingInterface, decouplin…

### PR DESCRIPTION
**Pull request by ruseinov**
Resolves https://github.com/paritytech/substrate/issues/12337 or rather what's left of it.
As we decided not to support pool joins at the moment we can probably ignore that requirement?

polkadot companion: https://github.com/paritytech/polkadot/pull/6199

### CODE REVIEW

File: primitives/staking/src/lib.rs [hunk 1]
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
 		num_slashing_spans: u32,
 	) -> Result<bool, DispatchError>;
 
+	/// Get StakingLedger by controller account.
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:138]:
```suggestion
	/// Get `StakingLedger` by controller account.
```

File: primitives/staking/src/lib.rs [hunk 2]
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
...
+	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
+
+	/// The ideal number of staking participants.
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:141]: The doc is very opaque, what is a staking participant? validator? nominator? active validator? validator candidate?

File: primitives/staking/src/lib.rs [hunk 3]
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
...
+	/// we do Staking::<T>::slashing_spans(&stash).iter().count() within it?
+	/// Or should we introduce another method force_unstake_with_spans that will do it?
+	fn force_unstake(
+		origin: OriginFor<Self>,
+		stash: T::AccountId,
+		num_slashing_spans: u32,
+	) -> DispatchResult;
+
+	/// Checks whether an account `staker` has been exposed in an era.
+	/// TODO: shall we have this helper here or just expose ErasStakers
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:160]: maybe something to keep in mind: https://github.com/paritytech/ci_cd/issues/626

File: primitives/staking/src/lib.rs [hunk 4]
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
...
+	/// storage instead for the caller to do as they please?
+	fn is_exposed_in_era(staker: &T::AccountId, era: &EraIndex) -> bool;
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:162]: I think for now this API is fine.

Comment (kianenigma) [reply primitives/staking/src/lib.rs:162]: Although -- I take it back. We have `get_ledger`. a similar `get_exposure` would fit nicer next to it.

File: primitives/staking/src/lib.rs (hunk 2 above)
Comment (ruseinov) [reply primitives/staking/src/lib.rs:141]: This doc is copied from the staking pallet though, gotta fix both.

File: primitives/staking/src/lib.rs (hunk 4 above)
Comment (ruseinov) [reply primitives/staking/src/lib.rs:153]: It's quite a bit different, because we have multiple exposures. 
So do we just return all the exposures that match staker, era? Otherwise it's inefficient.

I think it's best to keep this api tbh, because we are iterating by prefix. Or otherwise `get_exposures` with era.

File: primitives/staking/src/lib.rs (hunk 3 above)
Comment (ruseinov) [reply primitives/staking/src/lib.rs:151]: I try to keep it readable, but yeah, I guess those comments could be longer. Need to find a good tool for intellij

File: primitives/staking/src/lib.rs [hunk 5]
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
 		num_slashing_spans: u32,
 	) -> Result<bool, DispatchError>;
 
+	/// Get StakingLedger by controller account.
+	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
```
Comment (ruseinov) [inline primitives/staking/src/lib.rs:139]: @kianenigma We might want to get rid of those, seems redundant now.:
```
	fn active_stake(controller: &Self::AccountId) -> Option<Self::Balance> {
		Self::ledger(controller).map(|l| l.active)
	}

	fn total_stake(controller: &Self::AccountId) -> Option<Self::Balance> {
		Self::ledger(controller).map(|l| l.total)
	}
```

File: primitives/staking/src/lib.rs [hunk 6]
```diff
@@ -138,7 +138,7 @@ pub trait StakingInterface {
 	/// Get `StakingLedger` by controller account.
 	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
 
-	/// The ideal number of staking participants.
+	/// The ideal number of active validators.
```
Comment (ruseinov) [inline primitives/staking/src/lib.rs:141]: @kianenigma I think this reflects the situation better than before, judging by usages of this storage value.

File: frame/fast-unstake/src/lib.rs [hunk 7]
```diff
@@ -125,6 +125,9 @@ pub mod pallet {
 		/// The origin that can control this pallet.
 		type ControlOrigin: frame_support::traits::EnsureOrigin<Self::RuntimeOrigin>;
 
+		/// The access to staking functionality.
+		type Staking: StakingInterface<Balance = BalanceOf<Self>, AccountId = Self::AccountId>;
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:129]:
```suggestion
		type StakingInterface: StakingInterface<Balance = BalanceOf<Self>, AccountId = Self::AccountId>;
```

To be in-line with how `pallet-nomination-pools` is using this.

File: primitives/staking/src/lib.rs (hunk 5 above)
Comment (kianenigma) [reply primitives/staking/src/lib.rs:139]: Now I see why the old code was not returning `StakingLedger`. We cannot use `StakingLedger` in `pallet-nomination-pools` or `pallet-fast-unstake` because it is defined in pallet-stakin.

Putting it in a shared crate is also too much of a stretch in my opinion, because these types are arguably an implementation detail.

I think we should keep these, not have `get_ledger` and `get_exposure` and instead create other API function that express a condition, a state or a behavior rather than the explicit type.

File: primitives/staking/src/lib.rs (hunk 4 above)
Comment (kianenigma) [reply primitives/staking/src/lib.rs:153]: This is actually fine -- see my comment about `get_ledger`.

File: frame/fast-unstake/src/lib.rs (hunk 7 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:129]: I actually think that `type Staking` is a better name for it. StakingInterface seems redundant. Done this on purpose to float the idea.

File: primitives/staking/src/lib.rs (hunk 5 above)
Comment (ruseinov) [reply primitives/staking/src/lib.rs:139]: Why don't we introduce a special type that will take values out of StakingLedger? Like type `Ledger`. It could even be an alias for now and then change if needed. 

This makes very little sense to keep as is, because:
```
			let ledger =
				pallet_staking::Ledger::<T>::get(&ctrl).ok_or(Error::<T>::NotController)?;
			ensure!(!Queue::<T>::contains_key(&ledger.stash), Error::<T>::AlreadyQueued);
			ensure!(
				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
				Error::<T>::AlreadyHead
			);
			// second part of the && is defensive.
			ensure!(
				ledger.active == ledger.total && ledger.unlocking.is_empty(),
				Error::<T>::NotFullyBonded
			);
```

We'd need to call three different methods that operate on exactly the same storage item. Actually no, 4!
At the very least that's not very dry and also a bit of overhead.

File: primitives/staking/src/lib.rs (hunk 4 above)
Comment (ruseinov) [reply primitives/staking/src/lib.rs:153]: That's fine, we don't need eras anywhere else for now, so keeping this signature specific to the use-case sounds good and concise.

File: frame/fast-unstake/src/lib.rs (hunk 7 above)
Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:129]: I prefer keeping it consistent with `nomination-pools`. You can chose which. 

That being said, `type Foo: Foo` is a common pattern in Substrate and Rust in general.

Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:129]: Yeah, I guess I'm just not a big fan of the word `Interface` popping up in a type/variable name. It does make sense in a trait, because that's basically `Interface`.

Yep, I have been planning to make the nomination-pools change once the idea has been floated an accepted.

I also like `Staking`, because we will be able to pass `Staking` (pallet) to it and it'll be `Staking = Staking` :)

File: primitives/staking/src/lib.rs (hunk 5 above)
Comment (kianenigma) [reply primitives/staking/src/lib.rs:139]: my answer to this and many other aspects can be found in https://github.com/paritytech/substrate/compare/ru/feature/fast-unstake-istaking...kiz-ru/feature/fast-unstake-istaking

**Comment by ruseinov** [comment]
@kianenigma this needs your attention now. I've done fixes to nomination-pools and fast-unstake, only the tests/benches are failing now, but I'd like to verify the approach first.

**Comment by ruseinov** [comment]
bot rebase

File: frame/fast-unstake/Cargo.toml [hunk 8]
```diff
@@ -24,10 +24,6 @@ sp-io = { version = "6.0.0", default-features = false, path = "../../primitives/
 sp-runtime = { version = "6.0.0", default-features = false, path = "../../primitives/runtime" }
 sp-std = { version = "4.0.0", default-features = false, path = "../../primitives/std" }
 sp-staking = { default-features = false, path = "../../primitives/staking" }
-
-pallet-balances = { default-features = false, path = "../balances" }
-pallet-timestamp = { default-features = false, path = "../timestamp" }
```
Comment (kianenigma) [inline frame/fast-unstake/Cargo.toml:29]: oh boy 😮‍💨

File: frame/nomination-pools/src/mock.rs [hunk 9]
```diff
@@ -123,6 +111,48 @@ impl sp_staking::StakingInterface for StakingMock {
...
+		) {
+			(None, None) => Err(DispatchError::Other("balance not found")),
+			(Some(v), None) => Ok(Stake { total: v, active: 0, stash: who.clone() }),
+			(None, Some(v)) => Ok(Stake { total: v, active: v, stash: who.clone() }),
+			(Some(a), Some(b)) => Ok(Stake { total: a + b, active: b, stash: who.clone() }),
+		}
+	}
+
+	fn election_ongoing() -> bool {
+		todo!()
```
Comment (kianenigma) [inline frame/nomination-pools/src/mock.rs:132]:
```suggestion
		unimplemented!("reason?")
```

File: primitives/staking/src/lib.rs [hunk 10]
```diff
@@ -54,67 +56,93 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
 	}
 }
 
-/// Trait for communication with the staking pallet.
+/// A struct that reflects stake that an account has in the staking system. Provides a set of
+/// methods to operate on it's properties. Aimed at making `StakingInterface` more concise.
+pub struct Stake<T: StakingInterface + ?Sized> {
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:61]: In the staking pallet, a

`impl From<StakingLedger> for sp_staking::Stake` 

might come handy now.

**Review by kianenigma — COMMENTED** [review]
Looks overall good, will do a detailed review later. 

Has some leftover low hanging fruit nonetheless.

File: primitives/staking/src/lib.rs [hunk 11]
```diff
@@ -54,67 +56,93 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
...
-	/// The total stake that `stash` has in the staking system. This includes the
-	/// [`Self::active_stake`], and any funds currently in the process of unbonding via
-	/// [`Self::unbond`].
-	///
-	/// # Note
-	///
-	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
-	/// non-staking locks on the bonded pair's balance this may not be accurate.
-	fn total_stake(stash: &Self::AccountId) -> Option<Self::Balance>;
+	/// TODO: Possibly return a result here too for consistency.
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply primitives/staking/src/lib.rs:117]: @kianenigma I think returning a result here is kinda weird, but at the same time if there is some actual error and we just swallow it by returning a false - it might be a bad thing. WDYT?

File: primitives/staking/src/lib.rs (hunk 10 above)
Comment (ruseinov) [reply primitives/staking/src/lib.rs:61]: Can be done, though it's only going to be used in one method in staking pallet, so not sure if this is too helpful or more DRY than it is now

File: frame/nomination-pools/src/mock.rs (hunk 9 above)
Comment (ruseinov) [reply frame/nomination-pools/src/mock.rs:132]: Yep, basically we don't need those methods for nom-pools tests, I will make sure this is properly documented via reason.

File: frame/fast-unstake/Cargo.toml (hunk 8 above)
Comment (ruseinov) [reply frame/fast-unstake/Cargo.toml:29]: Yep, good stuff!

File: bin/node/runtime/src/lib.rs [hunk 12]
```diff
@@ -777,7 +779,7 @@ impl pallet_nomination_pools::Config for Runtime {
 	type RewardCounter = FixedU128;
 	type BalanceToU256 = BalanceToU256;
 	type U256ToBalance = U256ToBalance;
-	type StakingInterface = pallet_staking::Pallet<Self>;
+	type Staking = pallet_staking::Pallet<Self>;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:782]:
```suggestion
	type Staking = Staking; 
```

File: frame/fast-unstake/src/lib.rs [hunk 13]
```diff
@@ -125,6 +123,21 @@ pub mod pallet {
 		/// The origin that can control this pallet.
 		type ControlOrigin: frame_support::traits::EnsureOrigin<Self::RuntimeOrigin>;
 
+		/// Just the `Currency::Balance` type; we have this item to allow us to constrain it to
+		/// `From<u64>`.
+		type CurrencyBalance: sp_runtime::traits::AtLeast32BitUnsigned
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:128]: why is this needed?

File: frame/fast-unstake/src/lib.rs [hunk 14]
```diff
@@ -221,28 +234,23 @@ pub mod pallet {
 			let ctrl = ensure_signed(origin)?;
 
 			ensure!(ErasToCheckPerBlock::<T>::get() != 0, <Error<T>>::CallNotAllowed);
-
-			let ledger =
-				pallet_staking::Ledger::<T>::get(&ctrl).ok_or(Error::<T>::NotController)?;
-			ensure!(!Queue::<T>::contains_key(&ledger.stash), Error::<T>::AlreadyQueued);
+			let stash = T::Staking::stash_by_ctrl(&ctrl).map_err(|_| Error::<T>::NotController)?;
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:237]: for the sake of posterity, if we were to ignore controllers in this pallet and only allow stash to do operations, things would have been simpler. 

I.e. 

```
The dispatch origin of this call must be signed by the controller account, similar to
		/// `staking::unbond`.
```

is the root of evil. 

Perhaps we should already move to that direction, and either do it.

File: frame/fast-unstake/src/lib.rs [hunk 15]
```diff
@@ -221,28 +234,23 @@ pub mod pallet {
...
+			ensure!(!Queue::<T>::contains_key(&stash), Error::<T>::AlreadyQueued);
 			ensure!(
-				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
+				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != stash),
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:240]: isn't this shadowing the variable stash from the larger scope? 😱 if so, isn't there a compiler warning? why is no test failing? dangerous code. 

either way, please use a different name

File: frame/fast-unstake/src/lib.rs (hunk 14 above)
Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:237]: @rossbulat WDYT?

File: frame/nomination-pools/src/lib.rs [hunk 16]
```diff
@@ -622,17 +622,19 @@ impl<T: Config> BondedPool<T> {
 	///
 	/// This is often used for bonding and issuing new funds into the pool.
 	fn balance_to_point(&self, new_funds: BalanceOf<T>) -> BalanceOf<T> {
-		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+		let bonded_balance = T::Staking::stake(&self.bonded_account())
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:625]: I guess it is harmless to still have `fn active_balance` and `fn total_balance` on staking interface that are auto-implemented, preventing a little bit of dupe `.map`ing here.

File: primitives/staking/src/lib.rs [hunk 17]
```diff
@@ -19,8 +19,10 @@
 
 //! A crate which contains primitives that are useful for implementation that uses staking
 //! approaches in general. Definitions related to sessions, slashing, etc go here.
+extern crate core;
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:22]:
```suggestion
```

File: primitives/staking/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (kianenigma) [reply primitives/staking/src/lib.rs:117]: No strong opinion -- both are fine for me.

File: frame/fast-unstake/src/lib.rs (hunk 15 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:240]: It's not shadowing somehow, I thought it's a bit confusing, but looks kinda great :D Yes, will change the name.

File: bin/node/runtime/src/lib.rs (hunk 12 above)
Comment (ruseinov) [reply bin/node/runtime/src/lib.rs:782]: I was being consistent with the other Staking declarations in this node config. Should we then revisit those too?

File: frame/fast-unstake/src/lib.rs (hunk 13 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:128]: We actually don't need this anymore afaik, this is just leftover junk I ported over from Staking setup. Going to remove

File: frame/fast-unstake/src/lib.rs (hunk 14 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:237]: @kianenigma Probably best to do this move in both staking and fast-unstake to make the code less confusing?
Seeing how fast-unstake is already released with this interface anyway. We could make an issue for this though. 

What is the proposal here, passing the stash account directly?

File: frame/nomination-pools/src/lib.rs (hunk 16 above)
Comment (ruseinov) [reply frame/nomination-pools/src/lib.rs:625]: Yeah, I wanted to implement the extra stuff on `Stake` struct, but it'd still require mapping as we work with `Result`. So I guess we'd still be better off with bloating the interface a little.

File: frame/fast-unstake/src/lib.rs [hunk 18]
```diff
@@ -401,16 +406,9 @@ pub mod pallet {
...
-					stash.clone(),
-					num_slashing_spans,
-				);
+				let result = T::Staking::force_unstake(stash.clone());
 
-				let remaining = T::DepositCurrency::unreserve(&stash, deposit);
+				let remaining = T::Currency::unreserve(&stash, deposit);
 				if !remaining.is_zero() {
 					frame_support::defensive!("`not enough balance to unreserve`");
 					ErasToCheckPerBlock::<T>::put(0);
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:414]: You mean `halt`? Yeah, it makes sense, though since it's already implemented - just another conflict to solve.
I like it though, it's much more DRY and clear.

File: frame/fast-unstake/src/lib.rs (hunk 15 above)
Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:240]: would be curious to figure out exactly what is happening here, it seems important and I can see a similar mistake causing damage someday.

File: bin/node/runtime/src/lib.rs (hunk 12 above)
Comment (kianenigma) [reply bin/node/runtime/src/lib.rs:782]: No strong opinion, I think historically we use more the type alias rather than `Pallet` directly. 

Here, I suggested it for aesthetics. `X = X` looks better :D 

Personally, I dislike magic keyword that are not defined (`type Staking` is not visible in this scope), so I actually like yours better. 

Beyond the scope of this PR, these are all my thoughts on the matter, do as you prefer.

Comment (ruseinov) [reply bin/node/runtime/src/lib.rs:782]: fixed

File: frame/fast-unstake/src/lib.rs (hunk 15 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:240]: It's not a mistake. It works. Because `stash` in this case is a pattern that we match. So it works fine. We can't say `UnstakeRequest { stashAccount .. }` for example, because the field matches exactly.
And this is passing tests. So technically this is not a mistake, but it's confusing. So I will change the outer variable name.

Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs]: Scratch that, this should actually be shadowing the outer scope. But it does not warn.

Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs]: Yep, here's the deal:
1. It does shadow, so the equality comparison would always true
2. The reason it does not warn in this case is simply because the shadowing is only happening within `map_or` closure, so there's no `unused_variable` warning, because we use the original variable downstream.

Et voila!

File: frame/fast-unstake/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:406]: I guess I'd prefer this to remain in the scope of your batching PR to avoid confusion and duplication across the two. This can be done when this one is merged while rebasing into batching

**Comment by ruseinov** [comment]
@kianenigma I have addressed all the comments except for `halt` and the controller stuff.

I think this is outside of the scope of this PR (halt - because it's already implemented in another one, controller - because it's already released in the original version of fast-unstake, so it has little to do with the work we are doing here). 

Let me know if this makes sense.

File: frame/fast-unstake/src/lib.rs [hunk 19]
```diff
@@ -221,28 +222,25 @@ pub mod pallet {
...
+			ensure!(!Queue::<T>::contains_key(&stash_account), Error::<T>::AlreadyQueued);
 			ensure!(
-				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
+				Head::<T>::get()
+					.map_or(true, |UnstakeRequest { stash, .. }| stash_account != stash),
 				Error::<T>::AlreadyHead
 			);
-			// second part of the && is defensive.
-			ensure!(
-				ledger.active == ledger.total && ledger.unlocking.is_empty(),
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:234]: Nope, we are not exposing the ledger anymore, so the defensive condition has to go.

**Comment by ruseinov** [comment]
bot rebase

**Review by gilescope — APPROVED** [review]

File: frame/nomination-pools/src/lib.rs [hunk 20]
```diff
@@ -623,7 +623,7 @@ impl<T: Config> BondedPool<T> {
 	/// This is often used for bonding and issuing new funds into the pool.
 	fn balance_to_point(&self, new_funds: BalanceOf<T>) -> BalanceOf<T> {
 		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+			T::Staking::active_balance(&self.bonded_account()).unwrap_or(Zero::zero());
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:626]:
```suggestion
			T::Staking::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
```

File: frame/nomination-pools/src/lib.rs [hunk 21]
```diff
@@ -632,7 +632,7 @@ impl<T: Config> BondedPool<T> {
 	/// This is often used for unbonding.
 	fn points_to_balance(&self, points: BalanceOf<T>) -> BalanceOf<T> {
 		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+			T::Staking::active_balance(&self.bonded_account()).unwrap_or(Zero::zero());
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:635]:
```suggestion
			T::Staking::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
```

File: frame/nomination-pools/src/lib.rs [hunk 22]
```diff
@@ -738,7 +738,7 @@ impl<T: Config> BondedPool<T> {
 		ensure!(!self.is_destroying(), Error::<T>::CanNotChangeState);
 
 		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+			T::Staking::active_balance(&self.bonded_account()).unwrap_or(Zero::zero());
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:741]:
```suggestion
			T::Staking::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
```

File: frame/nomination-pools/src/lib.rs [hunk 23]
```diff
@@ -2212,7 +2188,7 @@ impl<T: Config> Pallet<T> {
 		debug_assert_eq!(frame_system::Pallet::<T>::consumers(&reward_account), 0);
 		debug_assert_eq!(frame_system::Pallet::<T>::consumers(&bonded_account), 0);
 		debug_assert_eq!(
-			T::StakingInterface::total_stake(&bonded_account).unwrap_or_default(),
+			T::Staking::total_balance(&bonded_account).unwrap_or_default(),
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:2191]:
```suggestion
			T::Staking::total_stake(&bonded_account).unwrap_or_default(),
```

File: frame/nomination-pools/src/tests.rs [hunk 24]
```diff
@@ -79,8 +79,8 @@ fn test_setup_works() {
 		let reward_account = Pools::create_reward_account(last_pool);
 
 		// the bonded_account should be bonded by the depositor's funds.
-		assert_eq!(StakingMock::active_stake(&bonded_account).unwrap(), 10);
-		assert_eq!(StakingMock::total_stake(&bonded_account).unwrap(), 10);
+		assert_eq!(StakingMock::active_balance(&bonded_account).unwrap(), 10);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:82]:
```suggestion
		assert_eq!(StakingMock::active_stake(&bonded_account).unwrap(), 10);
```

File: frame/nomination-pools/src/tests.rs [hunk 25]
```diff
@@ -79,8 +79,8 @@ fn test_setup_works() {
...
+		assert_eq!(StakingMock::total_balance(&bonded_account).unwrap(), 10);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:83]:
```suggestion
		assert_eq!(StakingMock::total_stake(&bonded_account).unwrap(), 10);
```

File: frame/nomination-pools/src/tests.rs [hunk 26]
```diff
@@ -2368,7 +2368,7 @@ mod unbond {
 				}
 			);
 
-			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
+			assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 0);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:2371]:
```suggestion
			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
```

File: frame/nomination-pools/src/tests.rs [hunk 27]
```diff
@@ -2415,7 +2415,7 @@ mod unbond {
 					]
 				);
 
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 94);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 94);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:2418]:
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 94);
```

File: frame/nomination-pools/src/tests.rs [hunk 28]
```diff
@@ -2443,7 +2443,7 @@ mod unbond {
 						}
 					}
 				);
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 2);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 2);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:2446]:
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 2);
```

File: frame/nomination-pools/src/tests.rs [hunk 29]
```diff
@@ -2486,7 +2486,7 @@ mod unbond {
 						}
 					}
 				);
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 0);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:2489]:
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
```

File: frame/nomination-pools/src/tests.rs [hunk 30]
```diff
@@ -2614,7 +2614,7 @@ mod unbond {
 						}
 					}
 				);
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 10);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 10);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:2617]:
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 10);
```

File: frame/nomination-pools/src/tests.rs [hunk 31]
```diff
@@ -2724,7 +2724,7 @@ mod unbond {
 					}
 				}
 			);
-			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
+			assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 0);
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:2727]:
```suggestion
			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
```

File: frame/nomination-pools/src/tests.rs [hunk 32]
```diff
@@ -3083,18 +3083,18 @@ mod pool_withdraw_unbonded {
...
 			assert_eq!(Balances::free_balance(&default_bonded_account()), 10);
 
 			// When
 			assert_ok!(Pools::pool_withdraw_unbonded(RuntimeOrigin::signed(10), 1, 0));
 
 			// Then there unbonding balance is no longer locked
-			assert_eq!(StakingMock::active_stake(&default_bonded_account()), Some(5));
-			assert_eq!(StakingMock::total_stake(&default_bonded_account()), Some(5));
+			assert_eq!(StakingMock::active_balance(&default_bonded_account()), Ok(5));
+			assert_eq!(StakingMock::total_balance(&default_bonded_account()), Ok(5));
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:3097]:
```suggestion
			assert_eq!(StakingMock::active_stake(&default_bonded_account()), Ok(5));
			assert_eq!(StakingMock::total_stake(&default_bonded_account()), Ok(5));
```

File: frame/nomination-pools/src/tests.rs [hunk 33]
```diff
@@ -3141,7 +3141,7 @@ mod withdraw_unbonded {
 					);
 					StakingMock::set_bonded_balance(
 						default_bonded_account(),
-						StakingMock::active_stake(&default_bonded_account()).unwrap() / 2,
+						StakingMock::active_balance(&default_bonded_account()).unwrap() / 2,
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:3144]:
```suggestion
						StakingMock::active_stake(&default_bonded_account()).unwrap() / 2,
```

File: frame/nomination-pools/src/tests.rs [hunk 34]
```diff
@@ -3270,7 +3270,7 @@ mod withdraw_unbonded {
 				// current bond is 600, we slash it all to 300.
 				StakingMock::set_bonded_balance(default_bonded_account(), 300);
 				Balances::make_free_balance_be(&default_bonded_account(), 300);
-				assert_eq!(StakingMock::total_stake(&default_bonded_account()), Some(300));
+				assert_eq!(StakingMock::total_balance(&default_bonded_account()), Ok(300));
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:3273]:
```suggestion
				assert_eq!(StakingMock::total_stake(&default_bonded_account()), Ok(300));
```

File: frame/nomination-pools/src/tests.rs [hunk 35]
```diff
@@ -4074,12 +4074,15 @@ mod create {
 			assert!(!BondedPools::<Runtime>::contains_key(2));
 			assert!(!RewardPools::<Runtime>::contains_key(2));
 			assert!(!PoolMembers::<Runtime>::contains_key(11));
-			assert_eq!(StakingMock::active_stake(&next_pool_stash), None);
+			assert_err!(
+				StakingMock::active_balance(&next_pool_stash),
```
Comment (kianenigma) [inline frame/nomination-pools/src/tests.rs:4078]:
```suggestion
				StakingMock::active_stake(&next_pool_stash),
```

File: primitives/staking/src/lib.rs [hunk 36]
```diff
@@ -54,67 +55,95 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
...
-		controller: Self::AccountId,
-		validators: sp_std::vec::Vec<Self::AccountId>,
-	) -> DispatchResult;
-
-	/// Chill `stash`.
-	fn chill(controller: Self::AccountId) -> DispatchResult;
-
-	/// Bond some extra amount in the _Stash_'s free balance against the active bonded balance of
-	/// the account. The amount extra actually bonded will never be more than the _Stash_'s free
+	fn active_balance(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:121]:
```suggestion
	fn active_stake(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```

File: primitives/staking/src/lib.rs [hunk 37]
```diff
@@ -54,67 +55,95 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
...
-	/// The total stake that `stash` has in the staking system. This includes the
-	/// [`Self::active_stake`], and any funds currently in the process of unbonding via
-	/// [`Self::unbond`].
-	///
-	/// # Note
-	///
-	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
-	/// non-staking locks on the bonded pair's balance this may not be accurate.
-	fn total_stake(stash: &Self::AccountId) -> Option<Self::Balance>;
+	fn total_balance(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```
Comment (kianenigma) [inline primitives/staking/src/lib.rs:117]:
```suggestion
	fn total_stake(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
renamed a few trivial things, LGTM otherwise.

**Comment by ruseinov** [comment]
bot merge

**Comment by ruseinov** [comment]
bot merge
